substrate/frame/contracts/src/benchmarking/call_builder.rs

```diff
@@ -0,0 +1,170 @@
+	<BalanceOf<T> as HasCompact>::Type: Clone + Eq + PartialEq + Debug + TypeInfo + Encode,
+{
+	/// Setup a new call for the given module.
+	pub fn new(module: WasmModule<T>) -> Self {
+		let contract = Contract::<T>::new(module.clone(), vec![]).unwrap();
+		let dest = contract.account_id.clone();
+		let origin = Origin::from_account_id(contract.caller.clone());
+
+		let storage_meter = Meter::new(&origin, None, 0u32.into()).unwrap();
+
+		Self {
+			contract,
+			dest,
+			origin,
+			gas_meter: GasMeter::new(Weight::MAX),
+			storage_meter,
+			schedule: T::Schedule::get(),
+			value: 0u32.into(),
+			debug_message: None,
+			determinism: Determinism::Enforced,
+			data: vec![],
+		}
+	}
+
+	/// Set the meter's storage deposit limit.
+	pub fn set_storage_deposit_limit(&mut self, balance: BalanceOf<T>) {
+		self.storage_meter = Meter::new(&self.origin, Some(balance), 0u32.into()).unwrap();
+	}
+
+	/// Set the call's origin.
+	pub fn set_origin(&mut self, origin: Origin<T>) {
+		self.origin = origin;
+	}
+
+	/// Set the contract's balance.
+	pub fn set_balance(&mut self, value: BalanceOf<T>) {
+		self.contract.set_balance(value);
+	}
+
+	/// Set the call's input data.
+	pub fn set_data(&mut self, value: Vec<u8>) {
+		self.data = value;
+	}
+
+	/// Set the debug message.
+	pub fn enable_debug_message(&mut self) {
+		self.debug_message = Some(Default::default());
+	}
+
+	/// Get the debug message.
+	pub fn debug_message(&self) -> Option<DebugBufferVec<T>> {
+		self.debug_message.clone()
+	}
+
+	/// Get the call's input data.
+	pub fn data(&self) -> Vec<u8> {
+		self.data.clone()
+	}
+
+	/// Get the call's contract.
+	pub fn contract(&self) -> Contract<T> {
+		self.contract.clone()
+	}
+
+	/// Build the call stack.
+	pub fn ext(&mut self) -> (StackExt<'_, T>, WasmBlob<T>) {
+		StackExt::bench_new_call(
+			self.dest.clone(),
+			self.origin.clone(),
+			&mut self.gas_meter,
+			&mut self.storage_meter,
+			&self.schedule,
+			self.value,
+			self.debug_message.as_mut(),
+			self.determinism,
+		)
+	}
+
+	/// Prepare a call to the module.
+	pub fn prepare_call<'a>(
+		ext: &'a mut StackExt<'a, T>,
+		module: WasmBlob<T>,
+		input: Vec<u8>,
+	) -> PreparedCall<'a, T> {
+		let (func, store) = module.bench_prepare_call(ext, input);
+		PreparedCall { func, store }
+	}
+}
+
+#[macro_export]
+macro_rules! call_builder(
+	($func: ident, $module:expr) => {
+		$crate::call_builder!($func, _contract, $module);
+	};
+	($func: ident, $contract: ident, $module:expr) => {
+		let mut setup = CallSetup::<T>::new($module);
+		$crate::call_builder!($func, $contract, setup: setup);
+	};
+    ($func:ident, setup: $setup: ident) => {
+		$crate::call_builder!($func, _contract, setup: $setup);
+	};
+    ($func:ident, $contract: ident, setup: $setup: ident) => {
+		let data = $setup.data();
+		let $contract = $setup.contract();
+		let (mut ext, module) = $setup.ext();
+		let $func = CallSetup::<T>::prepare_call(&mut ext, module, data);
+	};
```

Comment:
This will make the call side really hard to read because it just crates bindings which are not seen at the call side. I would prefer to return a tuple or a struct here.

Reply:
Not sure how I would do that, without raising some borrowing compilation errors.
$func borrow from ext that borrows from $setup, if I make it a function or try to return something, rustc will complain that something does not live long enough.

Reply:
I didn't realize you input the indentifiers so it is not actually as bad as I thought. My bad.

Reply:
mmm, to do so we would need to update 
```rust
pub struct Runtime<'a, E: Ext + 'a> {
	ext: &'a mut E,
	input_data: Option<Vec<u8>>,
	memory: Option<Memory>,
	chain_extension: Option<Box<<E::T as Config>::ChainExtension>>,
}
```
to take ownership of ext instead, is that what you had in mind ?

Reply:
That would be a big change and I am not sure if this would actually work. I meant just moving the `Ext` into the `PreparedCall` if that is easily possible.

Reply:
well then I don't see how we could do that easily without bothering the borrow checker.
Not too sure what that would buy us as well ...